### PR TITLE
feature/window frame groups

### PR DESCRIPTION
**Pull request by zembunia**
# Which issue does this PR close?

Closes #.

# Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/window/groups.rs [hunk 1]
```diff
@@ -46,12 +46,14 @@ impl WindowFrameGroups {
         if BISECT_SIDE || self.reached_end {
             return Ok(());
         }
-        let mut current_window_frame_end_idx = 0;
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/groups.rs]: Here I am proposing to use an exhaustive conditional instead with single static assignments instead of fall throughs and mutation for better readability. Can you verify that I am not breaking your logic?

File: datafusion/physical-expr/src/window/groups.rs [hunk 2]
```diff
@@ -74,15 +76,13 @@ impl WindowFrameGroups {
     ) -> Result<()> {
         if !BISECT_SIDE {
             return Ok(());
-        }
-        let mut initialize_start = false;
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/groups.rs]: In the same vein with my suggestions for the `extend_window_frame_if_necessary` function, I am trying to avoid mutation patterns here. The conditional is still not entirely exhaustive (the middle branch falls through), but it should be less prone to errors if were to make a change in the future. Can you please verify I am not breaking your logic?

File: datafusion/physical-expr/src/window/groups.rs [hunk 3]
```diff
@@ -152,12 +152,13 @@ impl WindowFrameGroups {
             }
         }
         if BISECT_SIDE {
-            let mut current_window_frame_start_idx = 0;
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/groups.rs]: In the same vein with my suggestions for the `extend_window_frame_if_necessary function`. Can you please verify I am not breaking your logic?

File: datafusion/physical-expr/src/window/groups.rs [hunk 4]
```diff
@@ -191,28 +192,23 @@ impl WindowFrameGroups {
             self.current_group_idx += 1;
             self.proceed_one_group::<BISECT_SIDE, SEARCH_SIDE>(item_columns, delta)?;
         }
-        if self.group_start_indices.is_empty() {
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/groups.rs]: I am suggesting to remove early returns, make the whole conditional a single exhaustive expression and use match patterns a little more effectively. None of these should break your code, but we should nevertheless that verify I am not breaking your logic?

File: datafusion/physical-expr/src/window/groups.rs [hunk 5]
```diff
@@ -46,12 +46,14 @@ impl WindowFrameGroups {
...
+        let current_window_frame_end_idx = if !SEARCH_SIDE {
+            self.current_group_idx + delta + 1
         } else if self.current_group_idx >= delta {
-            current_window_frame_end_idx = self.current_group_idx - delta + 1;
-        }
+            self.current_group_idx - delta + 1
+        } else {
+            0
+        };
+        // Can we remove the check below?
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/groups.rs:56]: I was not able to deduce if this check is necessary or not. Removing it does not break tests, but maybe the corner case it is guarding for is not getting exercised. Can we verify if this is needed, and if so, add a test to exercise it?

File: datafusion/physical-expr/src/window/groups.rs [hunk 6]
```diff
@@ -191,28 +192,23 @@ impl WindowFrameGroups {
...
-            let group_end_idx = match self.group_start_indices.back() {
-                Some(group_and_idx) => group_and_idx.1,
-                None => length,
-            };
-            Ok(group_end_idx)
-        }
+        })
     }
 
     fn find_next_group_and_start_index(
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/groups.rs:214]: This function should work even if the `step_size` is too small, or too large, right? Do we have test case exercising both of these situations?

**Review by ozankabak — COMMENTED** [review]
@zembunia, I checked in a few minor improvements. The changes are mainly about using exhaustivity in conditionals instead of fall-throughs, more succinct use of pattern matching and things like that. I am leaving in-line comments for you to verify that my suggestions do not break your logic.

Apart from my in-line comments and questions, I see only one general issue: We do not have any tests that exercise this implementation against built-in window functions such as `FIRST_VALUE`, `LAST_VALUE` etc. We should add some tests for such functions.

File: datafusion/physical-expr/src/window/groups.rs (hunk 5 above)
Comment (zembunia) [reply datafusion/physical-expr/src/window/groups.rs:56]: We can have the "delta PRECEDING" case for the end where the current group index is smaller than delta.

File: datafusion/physical-expr/src/window/groups.rs (hunk 2 above)
Comment (zembunia) [reply datafusion/physical-expr/src/window/groups.rs]: Do we still put an `else` after the returning first condition?
I think the middle branch should be ok without the condition, because it is the "delta FOLLOWING" case for start which makes it definitely greater than zero.

File: datafusion/physical-expr/src/window/groups.rs (hunk 5 above)
Comment (zembunia) [reply datafusion/physical-expr/src/window/groups.rs]: The case is covered in window.rs tests. The cases `SUM(c1) OVER (ORDER BY c3 GROUPS BETWEEN 9 PRECEDING AND 5 PRECEDING) as a2,
        SUM(c1) OVER (ORDER BY c3 GROUPS BETWEEN UNBOUNDED PRECEDING AND 5 PRECEDING) as a3` fail without the check.

File: datafusion/physical-expr/src/window/groups.rs [hunk 7]
```diff
@@ -78,9 +78,7 @@ impl WindowFrameGroups {
             return Ok(());
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/groups.rs:78]: An alternative structure that moves all fall-throughs to the end:

```rust
if BISECT_SIDE {
    if !SEARCH_SIDE {
        self.window_frame_start_idx = self.current_group_idx + delta;
        return self.initialize_window_frame_start(item_columns);
    } else if self.current_group_idx >= delta {
        self.window_frame_start_idx = self.current_group_idx - delta;
        return self.initialize_window_frame_start(item_columns);
    }
}
Ok(())
```

I leave it up to you to choose the current form, or this, depending on your personal taste.
